clippy_lints/src/derivable_impls.rs

```diff
@@ -81,13 +81,13 @@ fn check_struct<'tcx>(
     self_ty: &Ty<'_>,
     func_expr: &Expr<'_>,
     adt_def: AdtDef<'_>,
+    substs: SubstsRef<'_>,
 ) {
     if let TyKind::Path(QPath::Resolved(_, p)) = self_ty.kind {
-        if let Some(PathSegment { args: Some(a), .. }) = p.segments.last() {
-            for arg in a.args {
-                if !matches!(arg, GenericArg::Lifetime(_)) {
-                    return;
-                }
+        if let Some(PathSegment { args, .. }) = p.segments.last() {
+            let args = args.map(|a| a.args).unwrap_or(&[]);
+            if substs.len() != args.len() || args.iter().any(|arg| !matches!(arg, GenericArg::Lifetime(_))) {
```

Comment:
nit: comment explaining what this is checking (and why checking against `substs` solves the problem)

Reply:
Sure, will do while rebasing.

Reply:
This hasn't addressed my parenthetical:

Please make it clear why `substs.len() != args.len()` is sufficient to check that if there are defaults involved, so someone reading this code without a clear idea of `substs` can still understand

Reply:
Hopefully this is better now.